arch/x86/entry.S

```diff
@@ -69,6 +69,25 @@
 .macro syscall_to_usermode
     _to_usermode
 .endm
+
+.macro irq_from_usermode
+    testb $0x3, cpu_irq_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        _from_usermode
+        /* Preserve USER CS and a fake RIP on kernel stack */
+        /* This is expected by the irq_to_usermode         */
+        sub $cpu_irq_flags, %_ASM_SP
+        movw $__USER_CS, cpu_irq_cs(%_ASM_SP)
```

Comment:
Ugh! Can't we simply mimic a regular exception behaviour and push a zero error code and the IRQ's vector number to have a common stack frame (mod in-kernel IRQs and exceptions that lack the RSP+SS part but that's not really needed).